src/vm/amd64/pinvokestubs.S

```diff
@@ -0,0 +1,277 @@
+        lea rbp, [rsp + PInvokeStubForHostInner_FRAME_OFFSET]
+        END_PROLOGUE
+
+        mov             r10, rdx
+        mov             r12, rsi
+
+        test            rdi, rdi
+        jnz             LOCAL_LABEL(HandleStackArgs)
+
+        //
+        // Allocate space for scratch area if there are no stack args.
+        //
+        sub             rsp, PInvokeStubForHost_CALLEE_SCRATCH_SIZE
+
+LOCAL_LABEL(DoneStackArgs):
+        // unspill args
+        mov             rdi, [r12 + PInvokeStubForHost_PARAM_REGISTERS_OFFSET +  0h]
+        mov             rsi, [r12 + PInvokeStubForHost_PARAM_REGISTERS_OFFSET +  8h]
+        mov             rdx,  [r12 + PInvokeStubForHost_PARAM_REGISTERS_OFFSET + 10h]
+        mov             rcx,  [r12 + PInvokeStubForHost_PARAM_REGISTERS_OFFSET + 18h]
+        movdqa          xmm0, [r12 + PInvokeStubForHost_XMM_SAVE_OFFSET +  0h]
+        movdqa          xmm1, [r12 + PInvokeStubForHost_XMM_SAVE_OFFSET + 10h]
+        movdqa          xmm2, [r12 + PInvokeStubForHost_XMM_SAVE_OFFSET + 20h]
+        movdqa          xmm3, [r12 + PInvokeStubForHost_XMM_SAVE_OFFSET + 30h]
+        movdqa          xmm4, [r12 + PInvokeStubForHost_XMM_SAVE_OFFSET + 40h]
+        movdqa          xmm5, [r12 + PInvokeStubForHost_XMM_SAVE_OFFSET + 50h]
+        movdqa          xmm6, [r12 + PInvokeStubForHost_XMM_SAVE_OFFSET + 60h]
+        movdqa          xmm7, [r12 + PInvokeStubForHost_XMM_SAVE_OFFSET + 70h]
+
+        call            r10
+        
+        // spill return value
+        mov             [r12 + PInvokeStubForHost_XMM_SAVE_OFFSET +  0h], rax                                                
+        movdqa          [r12 + PInvokeStubForHost_XMM_SAVE_OFFSET + 10h], xmm0
+
+        // epilogue
+        lea             rsp, [rbp + PInvokeStubForHostInner_RETURN_ADDRESS_OFFSET - PInvokeStubForHostInner_NUM_REG_PUSHES*8]
+        pop             r12
+        pop             rbp
+        ret
+        
+// INPUTS: 
+//       rsi - number of stack bytes
+//       R12 - the outer method's frame pointer
+//       RSP -
+//       RBP - 
+//
+LOCAL_LABEL(HandleStackArgs):
+        //
+        // Allocate space for stack parameters  + scratch area.
+        //
+        sub             rsp, rdi
+        and             rsp, -16
+        sub             rsp, PInvokeStubForHost_CALLEE_SCRATCH_SIZE
+
+        //
+        // Copy stack parameters
+        //
+        shr             rdi, 3          // setup count
+       
+        mov             rcx, rdi
+        mov             rdx, rsi
+       
+        lea             rdi, [rsp + PInvokeStubForHost_CALLEE_SCRATCH_SIZE]      // rdi -> above callee scratch area
+        lea             rsi, [r12 + PInvokeStubForHost_PARAM_REGISTERS_OFFSET + PInvokeStubForHost_CALLEE_SCRATCH_SIZE]
+        rep             movsq
+        
+        mov             rsi, rdx             // restore rsi
+        mov             rdi, rcx             // restore rdi
```

Comment:
It seems we don't need to restore the rdi here, the windows version doesn't restore the rcx that corresponds to rdi. Anyways, this would zero the rdi since the rcx gets zeroed by the rep movsq